### ISSUE TITLE
Difference in RMSProp from published work

### ISSUE THREAD

**Issue by sergeyf** [fchollet/keras#435]
Howdy,

I was looking at this blog post http://sifter.org/~simon/journal/20150420.html in hopes of adding SMORMS3 to Keras, and discovered a small discrepancy between the way RMSProp is implemented in Keras and in published literature. Not sure if it's important, so posting here for the relevant eyeballs to look at.

In line 85 of optimizers.py, we have:

```new_p = p - self.lr * g / T.sqrt(new_a + self.epsilon)```

However, if we were to heed this recent paper (http://arxiv.org/pdf/1502.04390v1.pdf, equation after equation (2)), the code should be:

```new_p = p - self.lr * g / ( T.sqrt(new_a) + self.epsilon )```

In other words, the `self.epsilon` should not be inside the square root. 

This tutorial: https://github.com/Newmu/Theano-Tutorials/blob/master/4_modern_net.py also has the self.epsilon inside the square root, so not sure what the "standard" way is. The behavior would certainly be different...

Thoughts from someone who knows more about this than me?

PS, the RMSProp in this work by Alex Graves looks entirely different: http://arxiv.org/pdf/1308.0850v5.pdf

**Comment by fchollet** [comment]
You seem to misunderstand the role of epsilon here. It is mean to prevent numerical issues in discontinuous functions, while having no impact on the behavior of the underlying process. Which is why epsilon is chosen "small" (several orders of magnitude smaller than what you are interested in).

> In other words, the self.epsilon should not be inside the square root.

Here we need epsilon inside the sqrt becase taking sqrt of small values in Theano can trigger a NaN.

> This tutorial: https://github.com/Newmu/Theano-Tutorials/blob/master/4_modern_net.py also has the self.epsilon inside the square root, so not sure what the "standard" way is. The behavior would certainly be different...

No, it would not be different. Also, code you find on Github is not an appropriate measure of correctness.

**Comment by sergeyf** [comment]
I should have been clearer about what I meant by "the behavior would be different."

What I meant was that the default behavior would be different. If the epsilon is within the square root then the default epsilon of 1e-6 effectively becomes 1e-3 whereas for Adam,for example, the relevant code is

```(T.sqrt(v_b_t) + self.epsilon)```

The epsilon does not pass through the square root first.

I thought it would be good to have consistency as to whether self.epsilon passes through the square root across all of the optimizers. Especially if it doesn't matter to the behavior of the algorithms, as you point out. This would make it easier to do comparisons for fixed epsilons, making the default epsilons directly comparable.
